File: import-novels.js
const fs = require('fs');
const path = require('path');

// Novels folder path
const NOVELS_FOLDER = '/Users/abel/Desktop/Novels';

// Script.js path
const SCRIPT_PATH = path.join(__dirname, 'script.js');

// Function to read and process novel files
function importNovels() {
    console.log('Starting novel import process...');
    
    // Read files from novels folder
    const files = fs.readdirSync(NOVELS_FOLDER);
    
    // Filter TXT files
    const txtFiles = files.filter(file => file.endsWith('.txt'));
    
    console.log(`Found ${txtFiles.length} novel files.`);
    
    // Process each file
    const novels = {};
    
    txtFiles.forEach(file => {
        const filePath = path.join(NOVELS_FOLDER, file);
        const content = fs.readFileSync(filePath, 'utf8');
        
        // Extract novel info
        const { title, author, novelId } = extractNovelInfo(file);
        
        // Process content
        const processedContent = processNovelContent(content);
        
        // Add to novels object
        novels[novelId] = {
            title: title,
            author: author,
            content: processedContent
        };
        
        console.log(`Processed: ${title}`);
    });
    
    // Update script.js with new novels data
    updateScriptJs(novels);
    
    console.log('Novel import completed successfully!');
}

// Extract novel info from filename
function extractNovelInfo(filename) {
    // Remove .txt extension
    let name = filename.replace('.txt', '');
    
    // Extract title and author (simplified)
    let title = name;
    let author = 'Unknown Author';
    
    // Generate novel ID (convert to snake_case)
    const novelId = name.toLowerCase()
        .replace(/[\s-]/g, '_')
        .replace(/[\'’]/g, '')
        .replace(/[^a-z0-9_]/g, '');
    
    return { title, author, novelId };
}

// Process novel content
function processNovelContent(content) {
    // Split into paragraphs
    let paragraphs = content.split(/\n\s*\n/);
    
    // Clean paragraphs
    paragraphs = paragraphs
        .map(paragraph => paragraph.trim())
        .filter(paragraph => paragraph.length > 0)
        .slice(0, 20); // Take first 20 paragraphs for initial display
    
    return paragraphs;
}

// Update script.js with new novels data
function updateScriptJs(novels) {
    // Read current script.js
    let scriptContent = fs.readFileSync(SCRIPT_PATH, 'utf8');
    
    // Generate novels object string
    const novelsString = JSON.stringify(novels, null, 4);
    
    // Replace the novels object in script.js
    const updatedContent = scriptContent.replace(
        /const novels = \{[\s\S]*?\};/,
        `const novels = ${novelsString};`
    );
    
    // Write back to script.js
    fs.writeFileSync(SCRIPT_PATH, updatedContent, 'utf8');
    
    console.log('Updated script.js with new novels data.');
}

// Function to generate HTML for novel list items
function generateNovelListHTML() {
    console.log('Generating novel list HTML...');
    
    // Read files from novels folder
    const files = fs.readdirSync(NOVELS_FOLDER);
    const txtFiles = files.filter(file => file.endsWith('.txt'));
    
    let html = '';
    
    txtFiles.forEach(file => {
        const { title, author, novelId } = extractNovelInfo(file);
        
        html += `        <div class="novel-item" data-novel="${novelId}">
            <h3>${title}</h3>
            <p>${author}</p>
        </div>
`;
    });
    
    console.log('Generated novel list HTML:');
    console.log(html);
    
    return html;
}

// Main function
function main() {
    importNovels();
    generateNovelListHTML();
}

// Run if script is executed directly
if (require.main === module) {
    main();
}

module.exports = { importNovels, generateNovelListHTML };

File: import_novels.py
#!/usr/bin/env python3
"""
Import novels from txt files and convert to JavaScript format
"""
import os
import re
import json

# Novel files location
NOVELS_DIR = os.path.expanduser('~/Desktop/Novels')
OUTPUT_FILE = 'novels_data.js'

# Novel metadata (Chinese names)
NOVEL_META = {
    "Jane Eyre": {
        "author": "Charlotte Brontë",
        "chinese_name": "简·爱"
    },
    "Pride and Prejudice": {
        "author": "Jane Austen",
        "chinese_name": "傲慢与偏见"
    },
    "Moby Dick": {
        "author": "Herman Melville",
        "chinese_name": "白鲸"
    },
    "The Count of Monte Cristo": {
        "author": "Alexandre Dumas",
        "chinese_name": "基督山伯爵"
    },
    "Gulliver's Travels into Several Remote Nations of the World": {
        "author": "Jonathan Swift",
        "chinese_name": "格列佛游记",
        "short_name": "Gulliver's Travels"
    },
    "The Railway Children": {
        "author": "E. Nesbit",
        "chinese_name": "铁路少年"
    }
}

def clean_paragraph(text):
    """Clean and filter paragraphs"""
    # Remove excessive whitespace
    text = ' '.join(text.split())
    
    # Skip very short paragraphs (likely formatting artifacts)
    if len(text) < 50:
        return None
    
    # Skip common header/footer content
    skip_patterns = [
        r'Project Gutenberg',
        r'gutenberg\.org',
        r'END OF.*PROJECT GUTENBERG',
        r'START OF.*PROJECT GUTENBERG',
        r'\*\*\* (START|END)',
        r'Produced by',
        r'eBook',
        r'ebook',
    ]
    
    for pattern in skip_patterns:
        if re.search(pattern, text, re.IGNORECASE):
            return None
    
    return text

def process_novel(filepath):
    """Process a single novel file"""
    filename = os.path.basename(filepath)
    title = os.path.splitext(filename)[0]
    
    print(f"\n📚 Processing: {title}")
    
    # Get metadata
    meta = NOVEL_META.get(title, {
        "author": "Unknown Author",
        "chinese_name": title
    })
    
    # Read file
    with open(filepath, 'r', encoding='utf-8', errors='ignore') as f:
        content = f.read()
    
    # Split into paragraphs
    raw_paragraphs = content.split('\n')
    
    # Clean and filter paragraphs
    paragraphs = []
    for para in raw_paragraphs:
        cleaned = clean_paragraph(para)
        if cleaned:
            paragraphs.append(cleaned)
    
    print(f"   ✓ Extracted {len(paragraphs)} paragraphs")
    
    # Create novel ID (safe for JavaScript)
    novel_id = title.lower().replace("'", "").replace(" ", "_")
    novel_id = re.sub(r'[^a-z0-9_]', '', novel_id)
    
    return {
        'id': novel_id,
        'title': meta.get('short_name', title),
        'author': meta['author'],
        'chinese_name': meta['chinese_name'],
        'content': paragraphs[:2000]  # Limit to 2000 paragraphs per novel for performance
    }

def generate_javascript(novels):
    """Generate JavaScript file with novel data"""
    
    js_content = """// Novel Data - Auto-generated from txt files
// DO NOT EDIT MANUALLY - regenerate using import_novels.py

const NOVELS_DATA = {
"""
    
    for novel in novels:
        # Escape special characters in content
        content_json = json.dumps(novel['content'], ensure_ascii=False, indent=8)
        
        js_content += f"""    "{novel['id']}": {{
        "title": "{novel['title']}",
        "author": "{novel['author']}",
        "chineseName": "{novel['chinese_name']}",
        "content": {content_json}
    }},
"""
    
    js_content += """};

// Export for use in novels.js
if (typeof module !== 'undefined' && module.exports) {
    module.exports = NOVELS_DATA;
}
"""
    
    return js_content

def main():
    """Main function"""
    print("🔄 Importing novels from txt files...")
    print(f"📁 Source directory: {NOVELS_DIR}")
    
    # Check if directory exists
    if not os.path.exists(NOVELS_DIR):
        print(f"❌ Error: Directory not found: {NOVELS_DIR}")
        return
    
    # Process all txt files
    novels = []
    txt_files = [f for f in os.listdir(NOVELS_DIR) if f.endswith('.txt')]
    
    if not txt_files:
        print(f"❌ No .txt files found in {NOVELS_DIR}")
        return
    
    print(f"\n📖 Found {len(txt_files)} novel files")
    
    for filename in sorted(txt_files):
        filepath = os.path.join(NOVELS_DIR, filename)
        try:
            novel = process_novel(filepath)
            novels.append(novel)
        except Exception as e:
            print(f"   ❌ Error processing {filename}: {e}")
    
    # Generate JavaScript file
    print(f"\n💾 Generating {OUTPUT_FILE}...")
    js_content = generate_javascript(novels)
    
    with open(OUTPUT_FILE, 'w', encoding='utf-8') as f:
        f.write(js_content)
    
    print(f"✅ Successfully created {OUTPUT_FILE}")
    print(f"📊 Imported {len(novels)} novels:")
    for novel in novels:
        print(f"   - {novel['title']} ({novel['chinese_name']}): {len(novel['content'])} paragraphs")
    
    print("\n🎉 Import complete!")

if __name__ == '__main__':
    main()
